crates/wasmparser/src/primitives.rs

```diff
@@ -169,10 +170,65 @@ pub enum TypeDef<'a> {
     Module(ModuleType<'a>),
 }
 
-#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+/// A function type containing the function parameter and result types.
+///
+/// # Note
+///
+/// The parameters and results are ordered and merged in a single
+/// allocation starting with parameter types in order and following
+/// with the result types in order.
+#[derive(Clone, PartialEq, Eq, Hash)]
 pub struct FuncType {
-    pub params: Box<[Type]>,
-    pub returns: Box<[Type]>,
+    /// The ordered and merged parameters and results of the function type.
+    params_results: Box<[Type]>,
+    /// The number of inputs.
+    ///
+    /// The `len_params` field denotes how many inputs there are in
+    /// the head of the vector. The rest of the allocation is made up
+    /// of result types.
+    len_params: usize,
+}
+
+impl Debug for FuncType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FuncType")
+            .field("params", &self.params())
+            .field("returns", &self.returns())
+            .finish()
+    }
+}
+
+impl FuncType {
+    /// Creates a new function type.
+    pub fn new<I, O>(inputs: I, outputs: O) -> Self
+    where
+        I: IntoIterator<Item = Type>,
+        O: IntoIterator<Item = Type>,
+        I::IntoIter: ExactSizeIterator,
+    {
+        let inputs = inputs.into_iter();
+        let len_params = inputs.len();
+        let inputs_outputs = inputs.chain(outputs).collect::<Vec<_>>().into_boxed_slice();
+        Self {
+            params_results: inputs_outputs,
+            len_params,
+        }
+    }
+
+    /// Returns the parameter types of the function type.
+    pub fn params(&self) -> &[Type] {
+        &self.params_results[..self.len_params]
+    }
+
+    /// Returns the result types of the function type.
+    pub fn returns(&self) -> &[Type] {
+        &self.params_results[self.len_params..]
+    }
+
+    /// Returns the pair of parameter and result types of the function type.
+    pub fn params_returns(&self) -> (&[Type], &[Type]) {
+        self.params_results.split_at(self.len_params)
+    }
```

Comment:
I think this method can be omitted since the params/returns methods can be called together?

Reply:
We can remove this API indeed. We have this in `wasmi` because of some lifetime issues but there we additionally operate via a `Store` as done in Wasmtime. So maybe only the Wasmtime definition would probably benefit from this.